enable bf16 embeddingbag in aten

zhuhaozhe opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91949
* __->__ #89199

Summary:

This PR adds BF16 support For embedding_bag(CPU) op
 - Add bf16 in fp32 out embedding_lookup_idx kernel in Caffe2
 - Use Caffe2 kernel for BF16 fast path. Use FP32 buffer for accumulation (same as fp16 https://github.com/pytorch/pytorch/pull/74844)
 - Add UT coverage 
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.6721155643463135 | 0.6748666763305664  | 1.00  | TRUE        |
| 10       | 1.2210030555725098 | 1.211266279220581   | 0.99  | TRUE        |
| 20       | 2.381131649017334  | 1.7345335483551025  | 1.37  | TRUE        |
| 1        | 0.2621328830718994 | 0.31323862075805664 | 0.83  | FALSE       |
| 10       | 0.6128485202789307 | 0.6218626499176025  | 0.98  | FALSE       |
| 20       | 1.7913429737091064 | 1.2499897480010986  | 1.43  | FALSE       |

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

zhuhaozhe committed:
enable bf16 embeddingbag in aten

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable bf16 embeddingbag in aten"




[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable bf16 embeddingbag in aten"




[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -219,45 +245,52 @@ index_select_add(const Tensor &select_indices,
+      return;
+    }
+#endif
+#if !defined(USE_FBGEMM) or defined(BUILD_CAFFE2)
```
Do you need to check the data type is bfloat16 here?

zhuhaozhe replied:
> Do you need to check the data type is bfloat16 here?

Caffe2 support both fp16/bf16, so we do not need to check bf16 here. The difference is we can convert bf16->fp32 with vec but convert fp16->fp32 with scalar.

jgong5 replied:
But the code below explicitly uses BFloat16. Or, did I miss anything here? Also, there is also vec support for fp16<->fp32. This can be further improved in the future if we are using scalar conversion currently.

zhuhaozhe replied:
Only these two `using`s are explicitly related to bf16 but not check the data type. I think it's unnecessary.
```
    using bVec = vec::Vectorized<BFloat16>;
    using fVec = vec::Vectorized<float>;
```
Caffe2 kernels are same for bf16/fp16.
And fp32->bf16/fp16 part are different, I check the data type there.

For vec support fp16<->fp32 on CPU, I do not find them on cpu/vec folders. I will do some investigate in the future.

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update on "enable bf16 embeddingbag in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op
 - Add bf16 in fp32 out embedding_lookup_idx kernel in Caffe2
 - Use Caffe2 kernel for BF16 fast path. Use FP32 buffer for accumulation (same as fp16 https://github.com/pytorch/pytorch/pull/74844)
 - Add UT coverage 
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.6721155643463135 | 0.6748666763305664  | 1.00  | TRUE        |
| 10       | 1.2210030555725098 | 1.211266279220581   | 0.99  | TRUE        |
| 20       | 2.381131649017334  | 1.7345335483551025  | 1.37  | TRUE        |
| 1        | 0.2621328830718994 | 0.31323862075805664 | 0.83  | FALSE       |
| 10       | 0.6128485202789307 | 0.6218626499176025  | 0.98  | FALSE       |
| 20       | 1.7913429737091064 | 1.2499897480010986  | 1.43  | FALSE       |

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe commented:
Hi, @kit1980. May you help to review this too? It is the next PR to enable this kernel.https://github.com/pytorch/pytorch/pull/89198

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update on "enable bf16 embeddingbag in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op
 - Add bf16 in fp32 out embedding_lookup_idx kernel in Caffe2
 - Use Caffe2 kernel for BF16 fast path. Use FP32 buffer for accumulation (same as fp16 https://github.com/pytorch/pytorch/pull/74844)
 - Add UT coverage 
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.6721155643463135 | 0.6748666763305664  | 1.00  | TRUE        |
| 10       | 1.2210030555725098 | 1.211266279220581   | 0.99  | TRUE        |
| 20       | 2.381131649017334  | 1.7345335483551025  | 1.37  | TRUE        |
| 1        | 0.2621328830718994 | 0.31323862075805664 | 0.83  | FALSE       |
| 10       | 0.6128485202789307 | 0.6218626499176025  | 0.98  | FALSE       |
| 20       | 1.7913429737091064 | 1.2499897480010986  | 1.43  | FALSE       |

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable bf16 embeddingbag in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op
 - Add bf16 in fp32 out embedding_lookup_idx kernel in Caffe2
 - Use Caffe2 kernel for BF16 fast path. Use FP32 buffer for accumulation (same as fp16 https://github.com/pytorch/pytorch/pull/74844)
 - Add UT coverage 
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.6721155643463135 | 0.6748666763305664  | 1.00  | TRUE        |
| 10       | 1.2210030555725098 | 1.211266279220581   | 0.99  | TRUE        |
| 20       | 2.381131649017334  | 1.7345335483551025  | 1.37  | TRUE        |
| 1        | 0.2621328830718994 | 0.31323862075805664 | 0.83  | FALSE       |
| 10       | 0.6128485202789307 | 0.6218626499176025  | 0.98  | FALSE       |
| 20       | 1.7913429737091064 | 1.2499897480010986  | 1.43  | FALSE       |

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable bf16 embeddingbag in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op
 - Add bf16 in fp32 out embedding_lookup_idx kernel in Caffe2
 - Use Caffe2 kernel for BF16 fast path. Use FP32 buffer for accumulation (same as fp16 https://github.com/pytorch/pytorch/pull/74844)
 - Add UT coverage 
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.6721155643463135 | 0.6748666763305664  | 1.00  | TRUE        |
| 10       | 1.2210030555725098 | 1.211266279220581   | 0.99  | TRUE        |
| 20       | 2.381131649017334  | 1.7345335483551025  | 1.37  | TRUE        |
| 1        | 0.2621328830718994 | 0.31323862075805664 | 0.83  | FALSE       |
| 10       | 0.6128485202789307 | 0.6218626499176025  | 0.98  | FALSE       |
| 20       | 1.7913429737091064 | 1.2499897480010986  | 1.43  | FALSE       |

[ghstack-poisoned]

zhuhaozhe commented:
The failed test are not introduced by this PR.

zhuhaozhe committed:
Update on "enable bf16 embeddingbag in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op
 - Add bf16 in fp32 out embedding_lookup_idx kernel in Caffe2
 - Use Caffe2 kernel for BF16 fast path. Use FP32 buffer for accumulation (same as fp16 https://github.com/pytorch/pytorch/pull/74844)
 - Add UT coverage 
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.6721155643463135 | 0.6748666763305664  | 1.00  | TRUE        |
| 10       | 1.2210030555725098 | 1.211266279220581   | 0.99  | TRUE        |
| 20       | 2.381131649017334  | 1.7345335483551025  | 1.37  | TRUE        |
| 1        | 0.2621328830718994 | 0.31323862075805664 | 0.83  | FALSE       |
| 10       | 0.6128485202789307 | 0.6218626499176025  | 0.98  | FALSE       |
| 20       | 1.7913429737091064 | 1.2499897480010986  | 1.43  | FALSE       |

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable bf16 embeddingbag in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op
 - Add bf16 in fp32 out embedding_lookup_idx kernel in Caffe2
 - Use Caffe2 kernel for BF16 fast path. Use FP32 buffer for accumulation (same as fp16 https://github.com/pytorch/pytorch/pull/74844)
 - Add UT coverage 
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.6721155643463135 | 0.6748666763305664  | 1.00  | TRUE        |
| 10       | 1.2210030555725098 | 1.211266279220581   | 0.99  | TRUE        |
| 20       | 2.381131649017334  | 1.7345335483551025  | 1.37  | TRUE        |
| 1        | 0.2621328830718994 | 0.31323862075805664 | 0.83  | FALSE       |
| 10       | 0.6128485202789307 | 0.6218626499176025  | 0.98  | FALSE       |
| 20       | 1.7913429737091064 | 1.2499897480010986  | 1.43  | FALSE       |

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable bf16 embeddingbag in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op
 - Add bf16 in fp32 out embedding_lookup_idx kernel in Caffe2
 - Use Caffe2 kernel for BF16 fast path. Use FP32 buffer for accumulation (same as fp16 https://github.com/pytorch/pytorch/pull/74844)
 - Add UT coverage 
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.6721155643463135 | 0.6748666763305664  | 1.00  | TRUE        |
| 10       | 1.2210030555725098 | 1.211266279220581   | 0.99  | TRUE        |
| 20       | 2.381131649017334  | 1.7345335483551025  | 1.37  | TRUE        |
| 1        | 0.2621328830718994 | 0.31323862075805664 | 0.83  | FALSE       |
| 10       | 0.6128485202789307 | 0.6218626499176025  | 0.98  | FALSE       |
| 20       | 1.7913429737091064 | 1.2499897480010986  | 1.43  | FALSE       |

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe commented:
Hi, @jianyuh. May you help to review this PR?

jianyuh commented on CMakeLists.txt:
```diff
@@ -794,6 +794,10 @@ if(USE_COREML_DELEGATE)
   string(APPEND CMAKE_CXX_FLAGS " -DUSE_COREML_DELEGATE")
 endif()
 
+if(BUILD_CAFFE2)
```
This reverted in #91949 . We can avoid changing in this PR, or merge both PRs into one?

zhuhaozhe replied:
Sure, merged these two PRs into 1 https://github.com/pytorch/pytorch/pull/94163. Thanks for advice.

zhuhaozhe commented:
Close with https://github.com/pytorch/pytorch/pull/94163
